util/config.py

```diff
@@ -59,7 +53,22 @@ def initialize_globals():
     # Standard session configuration that'll be used for all new sessions.
     c.session_config = tf.ConfigProto(allow_soft_placement=True, log_device_placement=FLAGS.log_placement,
                                       inter_op_parallelism_threads=FLAGS.inter_op_parallelism_threads,
-                                      intra_op_parallelism_threads=FLAGS.intra_op_parallelism_threads)
+                                      intra_op_parallelism_threads=FLAGS.intra_op_parallelism_threads,
+                                      gpu_options=tf.GPUOptions(allow_growth=True))
+
+    # CPU device
+    c.cpu_device = '/cpu:0'
+
+    # Available GPU devices
+    c.available_devices = get_available_gpus(c.session_config)
```

Comment:
Isn't it easier to pass a single-purpose session config in this call rather than setting `c.session_config` twice? E.g.:

```python
    c.available_devices = get_available_gpus(
        tfv1.ConfigProto(gpu_options=tfv1.GPUOptions(allow_growth=True)))
```

And then leave the rest of the code in config.py unchanged.

Reply:
And could you write a small comment explaining why we have this special case?

Reply:
Yes, it's easier but I was suspicious after the entire memory allocation behavior, That's why I passed the full session configuration to prevent any unintended behavior. So if you are sure there is not such behavior, your solution will be sufficient.

I think this modification is essential in any case of running multiple concurrent trainings, otherwise it will not be possible because first training will allocate all of the available gpu memory.

> Isn't it easier to pass a single-purpose session config in this call rather than setting `c.session_config` twice? E.g.:
> 
> ```python
>     c.available_devices = get_available_gpus(
>         tfv1.ConfigProto(gpu_options=tfv1.GPUOptions(allow_growth=True)))
> ```
> 
> And then leave the rest of the code in config.py unchanged.

Yes, it's easier but I was suspicious after the entire memory allocation behavior, That's why I passed the full session configuration to prevent any unintended behavior. So if you are sure there is not such behavior, your solution will be sufficient.

Reply:
> And could you write a small comment explaining why we have this special case?

I think this modification is essential in any case of running multiple concurrent trainings, otherwise it will not be possible because first training will allocate all of the available gpu memory.

Reply:
Oh, I had misunderstood the changes due to how you described the problem in the PR. Changing the `get_available_gpus` call is not enough, you're setting that flag on the global session config that's also used for training. I thought the problem was listing the available GPUs specifically, but it's more general than that.